HTTP.markdown

```diff
@@ -0,0 +1,112 @@
+# Http
+
+If we have to following `.proto` file:
+
+```protobuf
+syntax = "proto3";
+
+import "google/api/annotations.proto";
+
+service Messaging {
+  rpc GetMessage(GetMessageRequest) returns (GetMessageResponse) {
+    option (google.api.http) = {
+      get:"/v1/messages/{message_id}"
+    };
+  }
+  rpc CreateMessage(CreateMessageRequest) returns (CreateMessageResponse) {
+    option (google.api.http) = {
+      post:"/v1/messages/{message_id}"
+      body: "message"
+    };
+  }
+  rpc UpdateMessage(CreateMessageRequest) returns (CreateMessageResponse) {
+    option (google.api.http) = {
+      patch:"/v1/messages/{message_id}"
+      body: "*"
+    };
+  }
+}
+
+message GetMessageRequest {
+  string message_id = 1;
+}
+message GetMessageResponse {
+  string message = 1;
+}
+
+message CreateMessageRequest {
+  string message_id = 1;
+  string message = 2; // mapped to the body
+}
+
+message CreateMessageResponse {}
+```
+
+The output will be
+
+```typescript
+// ...
+export interface GetMessageRequest {
+  messageId: string;
+}
+
+export interface GetMessageResponse {
+  message: string;
+}
+
+export interface CreateMessageRequest {
+  messageId: string;
+  /** mapped to the body */
+  message: string;
+}
+
+export interface CreateMessageResponse {}
+
+export const Messaging = {
+  getMessage: {
+    path: "/v1/messages/{message_id}",
+    method: "get",
+    requestType: undefined as GetMessageRequest,
+    responseType: undefined as GetMessageResponse,
+  },
+
+  createMessage: {
+    path: "/v1/messages/{message_id}",
+    method: "post",
+    body: "message",
+    requestType: undefined as CreateMessageRequest,
+    responseType: undefined as CreateMessageResponse,
+  },
+
+  updateMessage: {
+    path: "/v1/messages/{message_id}",
+    method: "patch",
+    body: "*",
+    requestType: undefined as CreateMessageRequest,
+    responseType: undefined as CreateMessageResponse,
+  },
+};
+```
+
+## Client implementation example
+
+```typescript
+// This is just an example, do not use it directly.
```

Comment:
Hey @jas-chen , I think this is a great so far--I'm wondering if we could further and remove the need for this `createApi` method...

I.e. if a user has ~10 services with ~10 methods each, they''ll have to call `createApi` against each of them--granted, they could write a method to do that in a loop...

Like our current non-google impl looks like:

```
export interface MyService {
  MyMethod(request: RequestType): Promise<ResponseType>;
}

export const MyServiceServiceName = "MyService";
export class MyServiceClientImpl implements MyService {
  private readonly rpc: Rpc;
  private readonly service: string;
  constructor(rpc: Rpc, opts?: { service?: string }) {
    this.service = opts?.service || MyServiceServiceName;
    this.rpc = rpc;
    this.MyMethod = this.MyMethod.bind(this);
  }
  MyMethod(request: RequestType): Promise<ResponseType> {
    const data = RequestType.encode(request).finish();
    const promise = this.rpc.request(this.service, "MyMethod", data);
    return promise.then((data) => ResponseType.decode(_m0.Reader.create(data)));
  }
}
```

And it seems like it'd be nice to provide a similar sort of API to users...

I suppose with the `export const Messaging`, the users could provide their own mapped types & helper methods to achieve this same output, i.e. something like:

```
// creates the `MessagingService.MyMethod` / `MessagingService.updateMessage` interface
type MessagingService = ClientService<typeof Messaging>;

const client: MessagingService = createClient(Messaging);
await client.getMessage(...);
```

I'm thinking it'd be great for ts-proto to eventually provide those `ClientService` and `createClient` implementations (i.e. that could automatically parse & replace the `{message_id}` based on the rules in their [docs](https://cloud.google.com/endpoints/docs/grpc-service-config/reference/rpc/google.api#google.api.HttpRule)).

Granted, sometimes I've just codegen-d out the `export interface MyService { ... }` and `export class MyServiceImpl`, which was a very old-school approach code generation, but it's definitely doable to use TS mapped types & runtime libraries to achieve the same effect.

Wdyt about the goal of shipping a `ClientService` & `createClient` method that took in the `Messaging` config? Maybe it could be like a `ts-proto-google-http` runtime library, which didn't bloat the codegen output. Would you be interesting in contributing that as well as this config output?

Thanks!

Reply:
I updated the client implementation example code, the `createApi` takes a service definition now.